pkg/ingest/store/local.go

```diff
@@ -0,0 +1,165 @@
+package store
+
+import (
+	"context"
+	"crypto/md5" //nolint:gosec
+	"encoding/hex"
+	"encoding/json"
+	"io"
+	"io/fs"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	nanoid "github.com/matoous/go-nanoid/v2"
+	"golang.org/x/exp/slices"
+)
+
+const cacheDirName = "_lakefs_cache"
+
+type LocalWalker struct {
+	mark            Mark
+	allowedPrefixes []string
+	cacheLocation   string
+}
+
+func NewLocalWalker(blockStorePath string, allowedPrefixes []string) *LocalWalker {
+	var cacheLocation string
+	if blockStorePath != "" {
+		cacheLocation = filepath.Join(blockStorePath, cacheDirName)
+	}
+	return &LocalWalker{
+		allowedPrefixes: allowedPrefixes,
+		mark:            Mark{HasMore: true},
+		cacheLocation:   cacheLocation,
+	}
+}
+
+func (l *LocalWalker) Walk(_ context.Context, storageURI *url.URL, options WalkOptions, walkFn func(e ObjectStoreEntry) error) error {
+	if storageURI.Scheme != "local" {
+		return path.ErrBadPattern
+	}
+	root := path.Join(storageURI.Host, storageURI.Path)
+	if err := l.verifyAbsPath(root); err != nil {
+		return err
+	}
+
+	var entries []*ObjectStoreEntry
+	// verify and use cache - location is stored in continuation token
+	if options.ContinuationToken != "" && strings.HasPrefix(options.ContinuationToken, l.cacheLocation) {
+		cacheData, err := os.ReadFile(options.ContinuationToken)
+		if err == nil {
+			err = json.Unmarshal(cacheData, &entries)
+			if err != nil {
+				entries = nil
+			} else {
+				l.mark.ContinuationToken = options.ContinuationToken
+			}
+		}
+	}
+
+	// if needed scan all entries to import and calc etag
+	if entries == nil {
+		if err := filepath.Walk(root, func(p string, info fs.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			key := filepath.ToSlash(p)
+			if key < options.After {
+				return nil
+			}
+			if !info.Mode().IsRegular() {
+				return nil
+			}
+
+			// calculate etag for the file
+			f, err := os.Open(p)
+			if err != nil {
+				return err
+			}
+			defer func() { _ = f.Close() }()
+			hash := md5.New() //nolint:gosec
+			_, err = io.Copy(hash, f)
```

Comment:
It will be very painful: AFAIK POSIX does _not_ guarantee that files are returned in sorted order.  So filepath.WalkDir itself needs to sort the directory somewhere.

Reply:
Update the code to perform MD5 as part of each iteration - it will reduce the first call time, while we perform the hash during each range.